In [1]:
import numpy as np 
import pandas as pd 
import statsmodels.api as sm
import matplotlib.pyplot as plt
import seaborn as sns
sns.set()

In [2]:
raw_data = pd.read_csv('1.0_train_data_set.csv')

In [3]:
data = raw_data.copy()

In [4]:
data['Admitted'] = data['Admitted'].map({'Yes': 1, 'No':0})
data['Gender'] = data['Gender'].map({'Female' :1, 'Male': 0})
data.head()

SAT  Admitted  Gender
0  1363         0       0
1  1792         1       1
2  1954         1       1
3  1653         0       0
4  1593         0       0

In [5]:
y = data['Admitted']
x1 = data['SAT']

In [6]:
x = sm.add_constant(x1)
reg_log = sm.Logit(y,x)
results_log = reg_log.fit()
results_log.summary()

Optimization terminated successfully.
         Current function value: 0.137766
         Iterations 10


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:               Admitted   No. Observations:                  168
Model:                          Logit   Df Residuals:                      166
Method:                           MLE   Df Model:                            1
Date:                Wed, 02 Nov 2022   Pseudo R-squ.:                  0.7992
Time:                        12:45:51   Log-Likelihood:                -23.145
converged:                       True   LL-Null:                       -115.26
Covariance Type:            nonrobust   LLR p-value:                 5.805e-42
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const        -69.9128     15.737     -4.443      0.000    -100.756     -39.070
SAT            0.0420      0.009      4.454      0.000       0.024       0.060
==============================================================================

Possibly complete quasi-separation: A fraction 0.27 of observations can be
perfectly predicted. This might indicate that there is complete
quasi-separation. In this case some parameters will not be identified.
"""

$ \log({odds}) = -0.69.9128 + 0.042 (SAT) $

$\Delta_{odds} = e^{β_k}$
  
 $ \log({odds_2}) = -0.69.9128 + 0.042 (SAT_2) $
 
$ \log({odds_1}) = -0.69.9128 + 0.042 (SAT_1) $
  
$ \log(\frac{odds_2}{odds_1})  = 0.042 (SAT_2 - SAT_1) $

$ \log(\frac{odds_{SAT_2}}{odds_{SAT_1}}) = 0.042 $


$ odds_{SAT_2} =1.042 (odds_{SAT_1}) $

$ odds_{SAT_2} = 104.2\% (odds_{SAT_1}) $

In [7]:
print(np.exp(0.042)) 
print(np.exp(0.42))

1.0428944787507632
1.5219615556186337


In [8]:
llr_pvalue = round(5.805e-42,3)
llr_pvalue

0.0

#### Observations:
- The logit model according to the logit regression table is shown above. 
- The model is significant and the SAT variable is significant too. 
- LLR p-value is 0.00 so our model is statistically different from LL-Null. 
- We created 2 other equations one for $ \log(odds_1) $ with $(SAT_1)$ variable and one for $\log(odds_2)$ with $(SAT_2)$ variable.
- The logistic regression coefficient β associated with a predictor X is the expected change in log odds of having the outcome per unit change in X. So increasing the predictor by 1 unit (or going from 1 level to the next) multiplies the odds of having the outcome by $e^β$ according to the following equation $\Delta_{odds} = e^{β_k}$
- Let $SAT_2$ equal to 1 and $SAT_1$ equal to 0. This way we can see outcome for increasing the predictor by 1 unit. 
- If we take the exponential of both sides
- $ odds_{SAT{2}} =1.042(odds_{SAT_{1}}) $
- When the SAT score increase by 1, the odds of getting admitted increase by 4.2%.
- If the SAT score increase by 10, the odds of getting addmitted increase by 52%.

In [9]:
y = data['Admitted']
x1 = data['Gender']

In [10]:
x = sm.add_constant(x1)
reg_log = sm.Logit(y,x)
results_log = reg_log.fit()
results_log.summary()

Optimization terminated successfully.
         Current function value: 0.572260
         Iterations 5


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:               Admitted   No. Observations:                  168
Model:                          Logit   Df Residuals:                      166
Method:                           MLE   Df Model:                            1
Date:                Wed, 02 Nov 2022   Pseudo R-squ.:                  0.1659
Time:                        12:45:51   Log-Likelihood:                -96.140
converged:                       True   LL-Null:                       -115.26
Covariance Type:            nonrobust   LLR p-value:                 6.283e-10
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const         -0.6436      0.222     -2.901      0.004      -1.078      -0.209
Gender         2.0786      0.363      5.727      0.000       1.367       2.790
==============================================================================
"""

In [11]:
np.exp(2.0786)

7.993270498536442

In [12]:
llr_pvalue = round(6.283e-10,3)
llr_pvalue

0.0

$ \log({odds}) = -0.64 + 2.0786 (gender) $

$\Delta_{odds} = e^{β_k}$
  
 $ \log({odds_2}) = -0.64 + 2.0786 (gender_2) $
 
$ \log({odds_1}) = -0.64 + 2.0786 (gender_1) $
  
$ \log(\frac{odds_2}{odds_1}) = 2.0786 (gender_2 - gender_1) $

$ \log(\frac{odds_{female}}{odds_{male}}) = 2.0786 $


$ odds_{female} =7.99 (odds_{male}) $

#### Observations:
- The logit model according to the logit regression table is shown above. 
- The model is significant and the gender variable is significant too.
- LLR p-value is 0.00 so our model is statistically different from LL-Null. 
- We created 2 other equations one for $ \log(odds_1) $ with $(gender_1)$ variable and one for $\log(odds_2)$ with $(gender_2)$ variable.
-  $\Delta_{odds} = e^{β_k}$
- Let $gender_2$ equal to 1 (female) and $gender_1$ equal to 0 (male).
- If we take the exponential of both sides
- $ odds_{female} =7.99 (odds_{male}) $
- According to the equation the odds of a female to get admitted are 7.99 the odds of a male to get admitted.



In [13]:
y = data['Admitted']
x1 = data[['SAT', 'Gender']]

In [14]:
x = sm.add_constant(x1)
reg_log = sm.Logit(y,x)
results_log = reg_log.fit()
results_log.summary()

Optimization terminated successfully.
         Current function value: 0.120117
         Iterations 10


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:               Admitted   No. Observations:                  168
Model:                          Logit   Df Residuals:                      165
Method:                           MLE   Df Model:                            2
Date:                Wed, 02 Nov 2022   Pseudo R-squ.:                  0.8249
Time:                        12:45:51   Log-Likelihood:                -20.180
converged:                       True   LL-Null:                       -115.26
Covariance Type:            nonrobust   LLR p-value:                 5.118e-42
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const        -68.3489     16.454     -4.154      0.000    -100.598     -36.100
SAT            0.0406      0.010      4.129      0.000       0.021       0.060
Gender         1.9449      0.846      2.299      0.022       0.287       3.603
==============================================================================

Possibly complete quasi-separation: A fraction 0.27 of observations can be
perfectly predicted. This might indicate that there is complete
quasi-separation. In this case some parameters will not be identified.
"""

In [15]:
np.exp(1.9449)

6.992932526814459

In [16]:
llr_pvalue = round(5.118e-42,3)
llr_pvalue

0.0

### Observations:
- This regression has a much higher Log-Likelihood value than the previous one, meaning current model is a better version. 
- Gender variable is still significant but it is no longer 0.00. It is a less significant variable in this model than the previous one.
- The new coefficient of gender is 1.9449 and it's exponential is 6.99. Which indicates that given the same SAT score the odds of a female to get admitted are 6.99 the odds of a male to get admitted.
- In this particular university or degree it is easier for females to enter.
- $ \log(\frac{odds_{female}}{odds_{male}}) = 1.9449 $

In [17]:
np.set_printoptions(formatter = {'float' :lambda x: "{0:0.2f}".format(x)})
results_log.predict()

array([0.00, 1.00, 1.00, 0.23, 0.02, 0.99, 1.00, 1.00, 1.00, 0.01, 1.00,
       1.00, 0.76, 0.00, 0.60, 1.00, 0.11, 0.12, 0.51, 1.00, 1.00, 1.00,
       0.00, 0.01, 0.97, 1.00, 0.48, 0.99, 1.00, 0.99, 0.00, 0.83, 0.25,
       1.00, 1.00, 1.00, 0.31, 1.00, 0.23, 0.00, 0.02, 0.45, 1.00, 0.00,
       0.99, 0.00, 0.99, 0.00, 0.00, 0.01, 0.00, 1.00, 0.92, 0.02, 1.00,
       0.00, 0.37, 0.98, 0.12, 1.00, 0.00, 0.78, 1.00, 1.00, 0.98, 0.00,
       0.00, 0.00, 1.00, 0.00, 0.78, 0.12, 0.00, 0.99, 1.00, 1.00, 0.00,
       0.30, 1.00, 1.00, 0.00, 1.00, 1.00, 0.85, 1.00, 1.00, 0.00, 1.00,
       1.00, 0.89, 0.83, 0.00, 0.98, 0.97, 0.00, 1.00, 1.00, 0.03, 0.99,
       0.96, 1.00, 0.00, 1.00, 0.01, 0.01, 1.00, 1.00, 1.00, 0.00, 0.00,
       0.02, 0.33, 0.00, 1.00, 0.09, 0.00, 0.97, 0.00, 0.75, 1.00, 1.00,
       0.01, 0.01, 0.00, 1.00, 0.00, 0.99, 0.57, 0.54, 0.87, 0.83, 0.00,
       1.00, 0.00, 0.00, 0.00, 1.00, 0.04, 0.00, 0.01, 1.00, 0.99, 0.52,
       1.00, 1.00, 0.05, 0.00, 0.00, 0.00, 0.68, 1.

In [18]:
np.array(data['Admitted'])

array([0, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1,
       0, 0, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 1, 0, 1, 0, 0, 0, 1, 1, 0,
       1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 1, 0, 1, 1, 1, 1, 0,
       0, 0, 1, 0, 1, 1, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 0, 1, 1, 0, 1,
       1, 1, 0, 0, 1, 1, 0, 1, 1, 0, 1, 1, 1, 0, 1, 0, 0, 1, 1, 1, 0, 0,
       0, 0, 0, 1, 0, 0, 1, 0, 1, 1, 1, 0, 0, 0, 1, 0, 1, 0, 1, 1, 1, 0,
       1, 0, 0, 0, 1, 0, 0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 1, 1, 1, 1, 1,
       1, 0, 1, 1, 0, 1, 0, 1, 1, 1, 1, 0, 0, 0], dtype=int64)

In [19]:
results_log.pred_table()

array([[69.00, 5.00],
       [4.00, 90.00]])

In [20]:
cm_df = pd.DataFrame(results_log.pred_table())
cm_df.columns = ['Predicted 0', 'Predicted 1']
cm_df = cm_df.rename(index={0:'Actual 0', 1:'Actual 1'})
cm_df

Predicted 0  Predicted 1
Actual 0         69.0          5.0
Actual 1          4.0         90.0

### Confusion Matrix
- For 69 observations the model predicted 0 and the true value was 0
- For 90 observations the model predicted 1 and the true value was 1
- For 4 observations the model predicted 0 while the true value was 1
- For 5 observations the model predicted 1 and the true value was 0
- Accurancy of this train set is 94.6%

In [21]:
# accurancy of the model 
cm = np.array(cm_df)
accurancy_train = (cm[0,0] + cm[1,1])/cm.sum()
accurancy_train

0.9464285714285714

In [22]:
test = pd.read_csv('2.03.+Test+dataset.csv')
test.head()

SAT Admitted  Gender
0  1323       No    Male
1  1725      Yes  Female
2  1762      Yes  Female
3  1777      Yes    Male
4  1665       No    Male

In [23]:
test['Admitted'] = test['Admitted'].map({'Yes': 1, 'No': 0})
test['Gender'] = test['Gender'].map({'Female': 1, 'Male': 0})
test

SAT  Admitted  Gender
0   1323         0       0
1   1725         1       1
2   1762         1       1
3   1777         1       0
4   1665         0       0
5   1556         1       1
6   1731         1       1
7   1809         1       1
8   1930         1       1
9   1708         1       0
10  1970         1       1
11  2039         1       1
12  1681         1       1
13  1444         0       0
14  1726         0       0
15  1876         1       1
16  1485         0       0
17  1500         0       1
18  1900         1       0

In [24]:
test_actual = test['Admitted']
test_data = test.drop(['Admitted'],axis=1)
test_data = sm.add_constant(test_data)
test_data

const   SAT  Gender
0     1.0  1323       0
1     1.0  1725       1
2     1.0  1762       1
3     1.0  1777       0
4     1.0  1665       0
5     1.0  1556       1
6     1.0  1731       1
7     1.0  1809       1
8     1.0  1930       1
9     1.0  1708       0
10    1.0  1970       1
11    1.0  2039       1
12    1.0  1681       1
13    1.0  1444       0
14    1.0  1726       0
15    1.0  1876       1
16    1.0  1485       0
17    1.0  1500       1
18    1.0  1900       0

In [25]:
def confusion_matrix(data,actual_values,model):    
    
        pred_values = model.predict(data)       
        bins=np.array([0,0.5,1])       
        cm = np.histogram2d(actual_values, pred_values, bins=bins)[0]   
        accuracy = (cm[0,0]+cm[1,1])/cm.sum()       
        return cm, accuracy

In [26]:
cm = confusion_matrix(test_data,test_actual,results_log)
print('confusion matrix :')
print(cm[0])

print('accuracy : ' +str(cm[1]))


confusion matrix :
[[5.00 1.00]
 [1.00 12.00]]
accuracy : 0.8947368421052632


In [27]:
cm_df = pd.DataFrame(cm[0])
cm_df.columns = ['Predicted 0','Predicted 1']
cm_df = cm_df.rename(index={0: 'Actual 0',1:'Actual 1'})
cm_df

Predicted 0  Predicted 1
Actual 0          5.0          1.0
Actual 1          1.0         12.0

In [28]:
print ('Missclassification rate: '+str((2)/19))

Missclassification rate: 0.10526315789473684
